hypershift-operator/controllers/machinedeployment/utils/providerspec.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"context"
+	"fmt"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/v2/api/v1beta2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// RegionAnnotation is the fallback annotation for AWS region
+	RegionAnnotation = "capa.infrastructure.cluster.x-k8s.io/region"
+)
+
+// ResolveAWSMachineTemplate fetches the AWSMachineTemplate referenced by the MachineDeployment
+func ResolveAWSMachineTemplate(ctx context.Context, c client.Client, machineDeployment *clusterv1.MachineDeployment) (*infrav1.AWSMachineTemplate, error) {
+	// Extract infrastructureRef
```

Comment:
We should perform both of these checks (infrastructureRef.name and infraRef.kind) before attempting to fetch the namespace of the machinedeployments in the main controller body.